Content.Client/UserInterface/Systems/Ghost/Controls/Roles/GhostRolesEui.cs

```diff
@@ -77,35 +76,45 @@ public override void HandleState(EuiStateBase state)
 
             if (state is not GhostRolesEuiState ghostState)
                 return;
+
+            // We must save BodyVisible state, so all Collapsible boxes will not close
+            // on adding new ghost role.
+            // Save the current state of each Collapsible box being visible or not
+            var collapsibleStates = _window.SaveCollapsibleBoxesStates();
+
+            // Clearing the container before adding new roles
             _window.ClearEntries();
 
             var entityManager = IoCManager.Resolve<IEntityManager>();
             var sysManager = entityManager.EntitySysManager;
             var spriteSystem = sysManager.GetEntitySystem<SpriteSystem>();
             var requirementsManager = IoCManager.Resolve<JobRequirementsManager>();
 
+            // Grouping roles
             var groupedRoles = ghostState.GhostRoles.GroupBy(
                 role => (role.Name, role.Description, role.Requirements));
+
+            // Add a new entry for each role group
             foreach (var group in groupedRoles)
             {
                 var name = group.Key.Name;
                 var description = group.Key.Description;
-                bool hasAccess = true;
-                FormattedMessage? reason;
-
-                if (!requirementsManager.CheckRoleRequirements(group.Key.Requirements, null, out reason))
-                {
-                    hasAccess = false;
-                }
+                var hasAccess = requirementsManager.CheckRoleRequirements(
+                    group.Key.Requirements,
+                    null,
+                    out var reason);
 
+                // Adding a new role
                 _window.AddEntry(name, description, hasAccess, reason, group, spriteSystem);
+
+                // Restore the Collapsible box state if it is saved
+                _window.RestoreCollapsibleBoxesStates(collapsibleStates);
```

Comment:
Shouldn't you only need to run this function once, after all the entries have been added?

Reply:
True. I forgot to move it out of the cycle